fxa-rust-client/src/login_sm.rs

```diff
@@ -0,0 +1,276 @@
+                } else if now > state.certificate_expires_at {
+                    info!("Certificate has expired. Transitioning to CohabitingAfterKeyPair.");
+                    CohabitingAfterKeyPair(state.token_keys_and_key_pair)
+                } else {
+                    info!("Key pair and certificate are fresh; staying Married.");
+                    Married(state) // same
+                }
+            }
+            CohabitingBeforeKeyPair(state) => {
+                debug!("Generating key pair.");
+                let key_pair = match FxAClient::key_pair(2048) {
+                    Ok(key_pair) => key_pair,
+                    Err(_) => {
+                        error!("Failed to generate key pair! Transitioning to Separated.");
+                        return Separated(state.base);
+                    }
+                };
+                info!("Key pair generated! Transitioning to CohabitingAfterKeyPairState.");
+                let new_state = CohabitingAfterKeyPairState {
+                    token_and_keys: state,
+                    key_pair,
+                    key_pair_expires_at: now() + 30 * 24 * 3600 * 1000,
+                };
+                CohabitingAfterKeyPair(new_state)
+            }
+            CohabitingAfterKeyPair(state) => {
+                debug!("Signing public key.");
+                let resp = self.client
+                    .sign(&state.token_and_keys.session_token, &state.key_pair);
+                match resp {
+                    Ok(resp) => {
+                        info!("Signed public key! Transitioning to Married.");
+                        let new_state = MarriedState {
+                            token_keys_and_key_pair: state,
+                            certificate: resp.certificate,
+                            certificate_expires_at: now() + 24 * 3600 * 1000,
+                        };
+                        Married(new_state)
+                    }
+                    Err(Error(err @ RemoteError(..), ..)) => {
+                        error!("Server error: {:?}. Transitioning to Separated.", err);
+                        Separated(state.token_and_keys.base)
+                    }
+                    Err(err @ _) => {
+                        error!(
+                            "Unknown error: ({:?}). Assuming transient, not transitioning.",
+                            err
+                        );
+                        CohabitingAfterKeyPair(state)
+                    }
+                }
+            }
+            EngagedBeforeVerified(state) => {
+                self.handle_ready_for_key_state(EngagedBeforeVerified, state)
+            }
+            EngagedAfterVerified(state) => {
+                self.handle_ready_for_key_state(EngagedAfterVerified, state)
+            }
+            Separated(_) => from,
+            Unknown => from,
+        }
+    }
+
+    fn handle_ready_for_key_state<F: FnOnce(ReadyForKeysState) -> FxALoginState>(
```

Comment:
A little comment about how you're using `same` would help.  And this construction is weird, since your "same" function really only ever refers to one type...